bika/lims/browser/analysisrequest/workflow.py

```diff
@@ -120,188 +116,92 @@ def notify_ar_retract(self, ar, newar):
-            hidden = True if hidden == 'on' else False
-            outs.append({'uid':uid, 'hidden':hidden})
+        hidden_ans = form.get('Hidden', {})
+        for uid in objects.keys():
+            hidden = hidden_ans.get(uid, '') == "on" or False
+            outs.append({'uid': uid, 'hidden': hidden})
         ar.setAnalysisServicesSettings(outs)
 
         specs = {}
-        if form.get("min", None):
-            for service_uid in Analyses:
-                service = objects[service_uid]
-                keyword = service.getKeyword()
-                specs[service_uid] = {
+        for service_uid, service in objects.items():
+            keyword = service.getKeyword()
+            results_range = ResultsRangeDict(keyword=keyword, uid=service_uid)
+            if form.get("min", None):
+                results_range.update({
                     "min": form["min"][0][service_uid],
                     "max": form["max"][0][service_uid],
                     "warn_min": form["warn_min"][0][service_uid],
                     "warn_max": form["warn_max"][0][service_uid],
-                    "keyword": keyword,
-                    "uid": service_uid,
-                }
-        else:
-            for service_uid in Analyses:
-                service = objects[service_uid]
-                keyword = service.getKeyword()
-                specs[service_uid] = ResultsRangeDict(keyword=keyword,
-                                                      uid=service_uid)
-        new = ar.setAnalyses(Analyses, prices=prices, specs=specs.values())
-        # link analyses and partitions
-        # If Bika Setup > Analyses > 'Display individual sample
-        # partitions' is checked, no Partitions available.
-        # https://github.com/bikalabs/Bika-LIMS/issues/1030
-        if 'Partition' in form:
-            for service_uid, service in objects.items():
-                part_id = form['Partition'][0][service_uid]
-                part = sample[part_id]
-                analysis = ar[service.getKeyword()]
-                analysis.setSamplePartition(part)
-                analysis.reindexObject()
-                partans = part.getAnalyses()
-                partans.append(analysis)
-                part.setAnalyses(partans)
-                part.reindexObject()
+                })
+            specs[service_uid] = results_range
 
-        if new:
-            if wasTransitionPerformed(ar, 'to_be_verified'):
-                # Apply to AR only; we don't want this transition to cascade.
-                ar.REQUEST['workflow_skiplist'].append("retract all analyses")
-                workflow.doActionFor(ar, 'retract')
-                ar.REQUEST['workflow_skiplist'].remove("retract all analyses")
+        if ar.setAnalyses(objects_uids, prices=prices, specs=specs.values()):
+            doActionFor(ar, "rollback_to_receive")
+
+        # Reindex the analyses
+        for analysis in ar.objectValues("Analysis"):
```

Comment:
When you add analyses by using "Manage analyses" view, you eventually add the analyses directly inside the current AR, so there is no need to reindex the analyses from child partitions, cause analyses from partitions haven't changed.... oh wait... I recall now you can remove analyses from there too... ouch!. But anyhow, even if we allow to remove analyses from children and it works properly, new analyses will always be added in the current AR (the primary one), never inside partitions. To make the thing a bit more consistent, I suggest to only allow the removal of analyses that directly belong to the current AR, not from children, in this "Manage analyses" view (make the checkbox readonly in such cases?) What do you think?

Reply:
I thought at the first moment to disallow "Manage Analyses" at all for Partitions, but I think that's not feasible. Then I would do it as you suggested and make Analyses which are not contained directly inside the AR to be readonly/disabled. Adding a note for that so that we don't forget.

Reply:
:+1: